google-cloud-firestore/lib/google/cloud/firestore/watch/listener.rb

```diff
@@ -0,0 +1,297 @@
+                      )
+                    )
+
+                    if @inventory.current? && @inventory.changes?
+                      synchronize do
+                        send_callback @inventory.build_query_snapshot
+                      end
+                    end
+                  when :CURRENT
+                    # The targets reflect all changes committed before the targets
+                    # were added to the stream.
+                    #
+                    # This will be sent after or with a +read_time+ that is
+                    # greater than or equal to the time at which the targets were
+                    # added.
+                    #
+                    # Listeners can wait for this change if read-after-write
+                    # semantics are desired.
+
+                    @inventory.persist(
+                      response.target_change.resume_token,
+                      Convert.timestamp_to_time(
+                        response.target_change.read_time
+                      )
+                    )
+
+                    @inventory.current = true
+                  when :RESET
+                    # The targets have been reset, and a new initial state for the
+                    # targets will be returned in subsequent changes.
+                    #
+                    # After the initial state is complete, +CURRENT+ will be
+                    # returned even if the target was previously indicated to be
+                    # +CURRENT+.
+
+                    @inventory.reset
+                    raise RestartStream # Raise to restart the stream
+                  end
+                when :document_change
+                  # A {Google::Firestore::V1beta1::Document Document} has changed.
+
+                  if response.document_change.removed_target_ids.any?
+                    @inventory.delete response.document_change.document.name
+                  else
+                    @inventory.add response.document_change.document
+                  end
+                when :document_delete
+                  # A {Google::Firestore::V1beta1::Document Document} has been
+                  # deleted.
+
+                  @inventory.delete response.document_delete.document
+                when :document_remove
+                  # A {Google::Firestore::V1beta1::Document Document} has been
+                  # removed from a target (because it is no longer relevant to
+                  # that target).
+
+                  @inventory.delete response.document_remove.document
+                when :filter
+                  # A filter to apply to the set of documents previously returned
+                  # for the given target.
+                  #
+                  # Returned when documents may have been removed from the given
+                  # target, but the exact documents are unknown.
+
+                  if response.filter.count != @inventory.size
```

Comment:
This needs to be the current size with all pending changes applied.

Reply:
So apply all pending changes, compare the count, then move the applied pending changes back to pending?

Reply:
Yes, that would work.

Reply:
Okay, this has been updated to apply the pending changes when calculating the count.

Reply:
Looks good.